File: dp/cpp/singleton.cpp
#include <iostream>
#include <memory>
#include <mutex>
#include <thread>
#include <vector>

// Thread-safe Singleton using modern C++
class DatabaseConnection {
private:
    static std::unique_ptr<DatabaseConnection> instance;
    static std::once_flag init_flag;

    // Private constructor
    DatabaseConnection() : connection_string("database://localhost:5432") {
        std::cout << "Database connection established\n";
    }

    std::string connection_string;

public:
    // Delete copy constructor and assignment operator
    DatabaseConnection(const DatabaseConnection&) = delete;
    DatabaseConnection& operator=(const DatabaseConnection&) = delete;

    // Thread-safe getInstance method
    static DatabaseConnection& getInstance() {
        std::call_once(init_flag, []() {
            instance = std::unique_ptr<DatabaseConnection>(new DatabaseConnection());
        });
        return *instance;
    }

    void executeQuery(const std::string& query) {
        std::cout << "Executing query: " << query << " on " << connection_string << std::endl;
    }

    void setConnectionString(const std::string& conn_str) {
        connection_string = conn_str;
    }

    const std::string& getConnectionString() const {
        return connection_string;
    }
};

// Static member definitions
std::unique_ptr<DatabaseConnection> DatabaseConnection::instance = nullptr;
std::once_flag DatabaseConnection::init_flag;

// Alternative: Meyer's Singleton (simpler, C++11 thread-safe)
class Logger {
private:
    Logger() = default;

public:
    static Logger& getInstance() {
        static Logger instance; // Thread-safe in C++11 and later
        return instance;
    }

    // Delete copy constructor and assignment operator
    Logger(const Logger&) = delete;
    Logger& operator=(const Logger&) = delete;

    void log(const std::string& message) {
        std::cout << "[LOG] " << message << std::endl;
    }
};

void workerThread(int id) {
    auto& db = DatabaseConnection::getInstance();
    db.executeQuery("SELECT * FROM users WHERE id = " + std::to_string(id));

    auto& logger = Logger::getInstance();
    logger.log("Worker thread " + std::to_string(id) + " completed");
}

int main() {
    std::cout << "=== Singleton Pattern Demo ===" << std::endl;

    // Test DatabaseConnection singleton
    auto& db1 = DatabaseConnection::getInstance();
    auto& db2 = DatabaseConnection::getInstance();

    std::cout << "Are both instances the same? " << (&db1 == &db2 ? "Yes" : "No") << std::endl;

    db1.executeQuery("SELECT * FROM products");
    db2.setConnectionString("database://remote:5432");
    db1.executeQuery("SELECT * FROM orders");

    // Test thread safety
    std::cout << "\nTesting thread safety:" << std::endl;
    std::vector<std::thread> threads;

    for (int i = 1; i <= 5; ++i) {
        threads.emplace_back(workerThread, i);
    }

    for (auto& t : threads) {
        t.join();
    }

    // Test Logger singleton
    std::cout << "\nTesting Logger singleton:" << std::endl;
    auto& logger1 = Logger::getInstance();
    auto& logger2 = Logger::getInstance();

    std::cout << "Are both logger instances the same? " << (&logger1 == &logger2 ? "Yes" : "No") << std::endl;

    logger1.log("Application started");
    logger2.log("Application running");

    return 0;
}

File: dp/rust/src/singleton.rs
/// Singleton Pattern Implementation in Rust
///
/// The Singleton pattern ensures a class has only one instance and provides
/// global access to it. In Rust, we use lazy_static or std::sync::Once for
/// thread-safe singleton implementation.

use std::sync::{Arc, Mutex, OnceLock};

/// Thread-safe Singleton using std::sync::OnceLock
#[derive(Debug)]
pub struct Database {
    connection_string: String,
}

impl Database {
    fn new() -> Self {
        println!("Creating database connection...");
        Database {
            connection_string: "postgresql://localhost:5432/myapp".to_string(),
        }
    }

    pub fn query(&self, sql: &str) -> Result<Vec<String>, String> {
        println!("Executing query: {}", sql);
        Ok(vec![format!("Result for: {}", sql)])
    }

    pub fn connection_string(&self) -> &str {
        &self.connection_string
    }
}

static DATABASE_INSTANCE: OnceLock<Arc<Mutex<Database>>> = OnceLock::new();

pub fn get_database() -> Arc<Mutex<Database>> {
    DATABASE_INSTANCE.get_or_init(|| {
        Arc::new(Mutex::new(Database::new()))
    }).clone()
}

/// Alternative implementation using lazy_static (requires lazy_static crate)
/// This is more idiomatic but requires an external dependency
/*
use lazy_static::lazy_static;

lazy_static! {
    static ref DATABASE: Arc<Mutex<Database>> = Arc::new(Mutex::new(Database::new()));
}

pub fn get_database_lazy() -> Arc<Mutex<Database>> {
    DATABASE.clone()
}
*/

/// Configuration Singleton for application settings
#[derive(Debug)]
pub struct Config {
    debug_mode: bool,
    max_connections: u32,
    timeout_seconds: u64,
}

impl Config {
    fn new() -> Self {
        Config {
            debug_mode: true,
            max_connections: 100,
            timeout_seconds: 30,
        }
    }

    pub fn set_debug_mode(&mut self, enabled: bool) {
        self.debug_mode = enabled;
    }

    pub fn debug_mode(&self) -> bool {
        self.debug_mode
    }

    pub fn max_connections(&self) -> u32 {
        self.max_connections
    }

    pub fn timeout_seconds(&self) -> u64 {
        self.timeout_seconds
    }
}

static CONFIG_INSTANCE: OnceLock<Arc<Mutex<Config>>> = OnceLock::new();

pub fn get_config() -> Arc<Mutex<Config>> {
    CONFIG_INSTANCE.get_or_init(|| {
        Arc::new(Mutex::new(Config::new()))
    }).clone()
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::thread;

    #[test]
    fn test_singleton_same_instance() {
        let db1 = get_database();
        let db2 = get_database();

        // Both should point to the same instance
        assert!(Arc::ptr_eq(&db1, &db2));
    }

    #[test]
    fn test_singleton_thread_safety() {
        let handles: Vec<_> = (0..10)
            .map(|i| {
                thread::spawn(move || {
                    let db = get_database();
                    let guard = db.lock().unwrap();
                    guard.query(&format!("SELECT * FROM table_{}", i))
                })
            })
            .collect();

        for handle in handles {
            let result = handle.join().unwrap();
            assert!(result.is_ok());
        }
    }

    #[test]
    fn test_config_singleton() {
        let config1 = get_config();
        let config2 = get_config();

        assert!(Arc::ptr_eq(&config1, &config2));

        // Test mutation
        {
            let mut guard = config1.lock().unwrap();
            guard.set_debug_mode(false);
        }

        {
            let guard = config2.lock().unwrap();
            assert!(!guard.debug_mode());
        }
    }
}

/// Example usage and demonstration
pub fn demo() {
    println!("=== Singleton Pattern Demo ===");

    // Get database instances
    let db1 = get_database();
    let db2 = get_database();

    // Verify they're the same instance
    println!("Same instance: {}", Arc::ptr_eq(&db1, &db2));

    // Use the database
    {
        let db = db1.lock().unwrap();
        let _ = db.query("SELECT * FROM users");
        println!("Connection: {}", db.connection_string());
    }

    // Configuration example
    let config = get_config();
    {
        let mut cfg = config.lock().unwrap();
        println!("Debug mode: {}", cfg.debug_mode());
        cfg.set_debug_mode(false);
        println!("Debug mode after change: {}", cfg.debug_mode());
    }
}

fn main() {
    println!("=== Singleton Pattern Demo ===");

    // Test Thread-Safe Singleton
    println!("\n1. Thread-Safe Singleton:");
    let instance1 = get_database();
    let instance2 = get_database();

    println!("Instances are same: {}", Arc::ptr_eq(&instance1, &instance2));

    {
        let db = instance1.lock().unwrap();
        println!("Connection: {}", db.connection_string());
        let _ = db.query("SELECT * FROM users");
    }

    // Test Configuration Singleton
    println!("\n2. Configuration Singleton:");
    let config1 = get_config();
    let config2 = get_config();

    println!("Config instances are same: {}", Arc::ptr_eq(&config1, &config2));

    {
        let mut cfg = config1.lock().unwrap();
        println!("Debug mode: {}", cfg.debug_mode());
        cfg.set_debug_mode(false);
        println!("Debug mode after change: {}", cfg.debug_mode());
    }
}